### PR TITLE
C4 155 Mitigation

### PR DESCRIPTION
**Pull request by dapp-whisperer**
https://github.com/code-423n4/2023-10-badger-findings/issues/155

### CODE REVIEW

**Review by GalloDaSballo — COMMENTED** [review]
Check the invariants from Medusa:
https://gist.github.com/GalloDaSballo/f7ef21e18f0c36de2793d2f7a88d5203

File: packages/contracts/contracts/CdpManagerStorage.sol [hunk 1]
```diff
@@ -516,7 +516,6 @@ contract CdpManagerStorage is EbtcBase, ReentrancyGuard, ICdpManagerData, AuthNo
                 uint256 _perUnitError
             ) = _calcSyncedGlobalAccounting(_newIndex, _oldIndex);
             _takeSplitAndUpdateFeePerUnit(_feeTaken, _newFeePerUnit, _perUnitError);
-            _updateSystemSnapshotsExcludeCollRemainder(0);
```
Comment (rayeaster) [inline packages/contracts/contracts/CdpManagerStorage.sol:519]: This change looks like not the solution and should be reverted https://github.com/code-423n4/2023-10-badger-findings/issues/155#issuecomment-1831082736

**Comment by dapp-whisperer** [comment]
I'm not clear _how_ the removal of the snapshot fixes the precision issue in 155.

**Comment by GalloDaSballo** [comment]
@CodingNameKiki @rayeaster Should SL_05 (sorting by NICR) be fixed after this PR?
What is the Error Threshold that would be tollerable?

**Comment by GalloDaSballo** [comment]
@CodingNameKiki @rayeaster 

Please review this specific sequence
```solidity
⇾ [FAILED] Property Test: EchidnaTester.echidna_sorted_list_invariant_5()
Test for method "EchidnaTester.echidna_sorted_list_invariant_5()" failed after the following call sequence:
[Call Sequence]
1) EchidnaTester.openCdp(16197885815696368879720681653477338690355059549524354304240887819103932625910, 209) (block=47434, time=580370, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
2) EchidnaTester.openCdp(16197885815696368879720681653477338690355059549524354304240887819103932625910, 209) (block=94867, time=1160739, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
3) EchidnaTester.setEthPerShare(4737424871052165462567343556913648738078620766275360444075220128633451887691) (block=115242, time=1612298, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000030000)
4) EchidnaTester.setEthPerShare(4737424871052165462567343556913648738078620766275360444075220128633451887691) (block=135617, time=2063857, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000030000)
5) EchidnaTester.withdrawColl(1, 528117742564021316393271938428361066789996829083) (block=175776, time=2355417, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
6) EchidnaTester.setEthPerShare(97056408238157249804947318527517112967233460345516200710872440659556098645798) (block=179903, time=2923763, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000010000)
```

Which breaks `invariant_SL_05` in spite of a 1e8 precision threshold

**Comment by rayeaster** [comment]
> @CodingNameKiki @rayeaster
> 
> Please review this specific sequence
> 
> ```solidity
> ⇾ [FAILED] Property Test: EchidnaTester.echidna_sorted_list_invariant_5()
> Test for method "EchidnaTester.echidna_sorted_list_invariant_5()" failed after the following call sequence:
> [Call Sequence]
> 1) EchidnaTester.openCdp(16197885815696368879720681653477338690355059549524354304240887819103932625910, 209) (block=47434, time=580370, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
> 2) EchidnaTester.openCdp(16197885815696368879720681653477338690355059549524354304240887819103932625910, 209) (block=94867, time=1160739, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
> 3) EchidnaTester.setEthPerShare(4737424871052165462567343556913648738078620766275360444075220128633451887691) (block=115242, time=1612298, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000030000)
> 4) EchidnaTester.setEthPerShare(4737424871052165462567343556913648738078620766275360444075220128633451887691) (block=135617, time=2063857, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000030000)
> 5) EchidnaTester.withdrawColl(1, 528117742564021316393271938428361066789996829083) (block=175776, time=2355417, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000020000)
> 6) EchidnaTester.setEthPerShare(97056408238157249804947318527517112967233460345516200710872440659556098645798) (block=179903, time=2923763, gas=12500000, gasprice=1, value=0, sender=0x0000000000000000000000000000000000010000)
> ```
> 
> Which breaks `invariant_SL_05` in spite of a 1e8 precision threshold

Good news is that this sequence is caught by our own medusa/fuzzing run and it reflects exactly the same problem as [C4 #155](https://github.com/code-423n4/2023-10-badger-findings/issues/155) reported which we tend to [agree on "No Fix"](https://docs.google.com/document/d/1OGVCASm91gVsAXaT3C1uGdtBP1KIqsUZ2g_aaY77znE/edit)

The specific NICRs in above sequence are quite quite large (bigger than $9*10^{15}$) which means the affected CDPs are on top of the sorted list thus making them not eligible for liquidation or redemption in practice. 

I would recommend to modify the invariant properties `invariant_SL_05` as following (i.e.,we tolerate precision error by using the relative ratio of NICR difference, not the absolute difference value):

```diff
++  uint256 NICR_PRECISION = 1e20;
    function invariant_SL_05(CRLens crLens, SortedCdps sortedCdps) internal returns (bool) {
        bytes32 currentCdp = sortedCdps.getFirst();

        uint256 newIcrPrevious = type(uint256).max;

        while (currentCdp != bytes32(0)) {
--          uint256 newIcr = crLens.quoteRealICR(currentCdp);
++          uint256 newIcr = crLens.quoteRealNICR(currentCdp);
            if (newIcr > newIcrPrevious) {
                /// @audit Precision Threshold to flag very scary scenarios
                /// Innoquous scenario illustrated here: https://github.com/Badger-Finance/ebtc-fuzz-review/issues/15
--              if (newIcr - newIcrPrevious > NICR_ERROR_THRESHOLD) {
++              if (((newIcr - newIcrPrevious) * NICR_PRECISION / newIcrPrevious) > NICR_ERROR_THRESHOLD) {
                    return false;
                }
            }
            newIcrPrevious = newIcr;

            currentCdp = sortedCdps.getNext(currentCdp);
        }
        return true;
    }

```

**Pull request closed without merging**